[egs] icsi recipe

pswietojanski opened the pull request:
Hey,

I had parts of this recipe for some time, but never really had time to finish it properly, until recently. I added missing bits to make it fully run-able (given someone's access to ICSI data). One can build models for three conditions with some flexibility w.r.t what mics to use in sdm and mdm cases.

This PR includes results for GMM systems, at some point I will update this with mini-librispeech compatible chain models (but due to time constraints, this may take some time, so just committing baseline systems for now). Perhaps someone else will get to nnet3 baselines faster,  given data prep is there and reasonably tested.

Recipe is based on AMI s5b as much as possible. Let me know if there is something you would like to me to update, before you are happy with PR quality.

Best,
Pawel

danpovey commented:
Great-- thanks!
How does it relate to the different versions of the AMI scripts?
There might be a couple of unnecessary files there e.g. topo.proto and plp.conf.

pswietojanski commented:
conf/* is basically copied (except beamformer.cfg). The data preparation pipeline is entirely custom, to take into account all icsi data specifics. local/convert2stm.py and local/glm are the same (should I soft-link them?).

danpovey commented:
I prefer copies to soft links.
Let's wait and see if anyone who is familiar with the AMI setup (Vimal?
Shinji?) can comment.


On Sun, May 20, 2018 at 7:49 PM, Pawel Swietojanski <
notifications@github.com> wrote:

> conf/* is basically copied (except beamformer.cfg). The data preparation
> pipeline is entirely custom, to take into account all icsi data specifics.
> local/convert2stm.py and local/glm are the same (should I soft-link them?).
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2435#issuecomment-390522188>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ADJVu4-r55H6MeC0o6OcoG71CTflfvxBks5t0gEUgaJpZM4UGRfF>
> .
>

pswietojanski commented:
Thanks. So the reason I decided for LDA distribution in the end is that it offers complete corpora (including all distant channels). This is something that seems to be missing from Edinburgh download (only ihm). And I assumed most universities have LDC-released ICSI data already in their repositories. I do mention http://groups.inf.ed.ac.uk/ami/icsi/ in the README though, so ppl are aware at least ihm should be available from there.

pswietojanski commented:
Hey @danpovey so I had some time over the last weekend and added / trained chain models for each ICSI scenario (including ihm alignments for mdm and sdm setups). I guess this is ready to be merged at this point, I am not planning any further additions in the foreseeable future.

danpovey commented on egs/icsi/README.txt:
```diff
@@ -0,0 +1,99 @@
+
+To train models, for an arbitrary mic, go to s5 run something like (after you
```
can you please move the "instructional" type recipes to a comment at the top of the run.sh, and have this README.txt more about the data itself?  Please prominently say how many hours of train/test, what language, what type of speech (I assume meeting, with overlaps and multiple microphones), and what sampling rate.

danpovey replied:
... and also make clear prominently, at the top, that it's available from the LDC and mention the corpus numbers.
The idea is that people who are looking for setups with free data will know quickly not to look hdre.

pswietojanski commented:
Thanks for the feedback, updated as requested. Fail of the check build of the latest update to this PR seems independent of it, I guess some infrastructure error on Travis side.

jtrmal commented:
yes, I have even tried to restart the build and it failed again. I also
think it's infrastructure problem.
y.

On Tue, Jun 19, 2018 at 3:33 PM Pawel Swietojanski <notifications@github.com>
wrote:

> Thanks for the feedback, updated as requested. Fail of the check build of
> the latest update to this PR seems independent of it, I guess some
> infrastructure error on Travis side.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2435#issuecomment-398518431>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX6vbKwK8W2DycFptAWqldfmY6kmhks5t-VILgaJpZM4UGRfF>
> .
>

pswietojanski commented:
Should I do some dummy PR update to re-trigger the checks today, or one can do in some other way?

jtrmal commented:
I have restarted it.
y.

On Wed, Jun 20, 2018 at 11:58 AM Pawel Swietojanski <
notifications@github.com> wrote:

> Should I do some dummy PR update to re-trigger the checks today, or one
> can do in some other way?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/2435#issuecomment-398802848>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AKisX67lGVQNLk2Eu6lgKbl6CoNexVwEks5t-nEkgaJpZM4UGRfF>
> .
>

pswietojanski commented:
I just noticed you were not explicitly mentioned as the reviewers of this PR (thus perhaps overlooked this) @sw005320 @vimalmanohar Please could you have a look at your spare time? Cheers, Pawel

sw005320 commented:
Sure, I'll check some of them (possibly weekend).

vimalmanohar commented on egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh:
```diff
@@ -0,0 +1,266 @@
+stage=0
+mic=ihm
+nj=30
+min_seg_len=1.55
```
Recent scripts don't need min_seg_len and can instead provide multiple frames-per-chunk like 160,140,100 etc. The AMI recipe is also not updated with this change. But this is something that can be done in this recipe to reduce the many complexities involved in combing short segment feats.

vimalmanohar commented on egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh:
```diff
@@ -0,0 +1,266 @@
+gmm=tri3  # the gmm for the target data
+ihm_gmm=tri3  # the gmm for the IHM system (if --use-ihm-ali true).
+num_threads_ubm=3
+ivector_transform_type=pca
```
I think it is better to make the pca version as default and remove it as an option.

vimalmanohar commented on egs/icsi/s5/local/icsi_beamform.sh:
```diff
@@ -0,0 +1,101 @@
+
+if [ $# != 3 ]; then
+   echo "Wrong #arguments ($#, expected 4)"
+   echo "Usage: steps/ami_beamform.sh [options] <num-mics> <ami-dir> <wav-out-dir>"
```
ami -> icsi

vimalmanohar commented on egs/icsi/s5/local/nnet3/prepare_lores_feats.sh:
```diff
@@ -0,0 +1,84 @@
+  utils/fix_data_dir.sh data/${mic}/${train_set}${ihm_suffix}_sp
+fi
+
+#we do not combine segments here by default, as have not done experiment if it helps
```
It seems like min_seg_len is not even used. It is better to remove this and try instead to provide multiple frame-chunks.

sw005320 commented on egs/icsi/s5/conf/beamformit.cfg:
```diff
@@ -0,0 +1,56 @@
+#BeamformIt sample configuration file for AMI data (http://groups.inf.ed.ac.uk/ami/download/)
```
AMI -> ICSI
Also, remove the URL

sw005320 commented on egs/icsi/s5/local/chain/tuning/run_tdnn_1a.sh:
```diff
@@ -0,0 +1,266 @@
+if [ $stage -le 16 ]; then
+  if [[ $(hostname -f) == *.clsp.jhu.edu ]] && [ ! -d $dir/egs/storage ]; then
+    utils/create_split_dir.pl \
+     /export/b0{5,6,7,8}/$USER/kaldi-data/egs/ami-$(date +'%m_%d_%H_%M')/s5b/$dir/egs/storage $dir/egs/storage
```
ami->icsi

sw005320 commented on egs/icsi/s5/local/icsi_agree_words_replace.pl:
```diff
@@ -0,0 +1,34 @@
+#!/usr/bin/perl
+
+if (@ARGV != 3) {
+  print STDERR "Usage: ami_agree_words.pl <replace-file> <in-segments> <out-segments>\n";
```
ami->icsi

sw005320 commented on egs/icsi/s5/local/icsi_agree_words_split.pl:
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/perl
+
+if (@ARGV != 3) {
+  print STDERR "Usage: ami_agree_words_split.pl <split-file> <in-segments> <out-segments>\n";
```
ami->icsi

sw005320 commented on egs/icsi/s5/run.sh:
```diff
@@ -0,0 +1,172 @@
+
+set -euo pipefail
+
+# Path where AMI gets downloaded (or where locally available):
```
These comments look like for AMI

sw005320 commented on egs/icsi/s5/run.sh:
```diff
@@ -0,0 +1,172 @@
+  utils/subset_data_dir.sh data/$mic/train 15000 data/$mic/train_15k
+
+  steps/train_mono.sh --nj $nj --cmd "$train_cmd" \
+    data/$mic/train_15k data/lang exp/$mic/mono
```
It depends on the task, and amount of training data, but I usually observed that training with original noisy speech is better than that with enhanced speech (or using both), when we use DNN acoustic models. Using all microphone data is also good in terms of we'll increase the amount of speech data. The chime5 recipe is similar except for this part.

danpovey replied:
@sw005320, can you please clarify that comment above?
Were you asking for them to make some specific change in the recipe? 
Are you saying that the enhancement is not helpful for training but can still help for test?

sw005320 replied:
> Are you saying that the enhancement is not helpful for training but can still help for test?
Right.
But this needs experiments, as it is not always true. This is just a comment.

pswietojanski commented:
Thanks for the feedback, cleaned some top level bits (mostly typos) and removed variables that were not used anyway (but were still declared as overlooked this). At this point will not change a default chunk_width (from 150), as would need to re-train chain models for that, and do not have capacity to do so. I would expect this to be a less of an issue for ICSI (when compared to AMI), as there are fewer shorter segments overall (below 1.5s) but anyway, it is substantial so will have a look at this, when re-training chain models will be an option on my side.

pswietojanski commented:
An update, re-ran all mic conditions with frames_per_eg set to 150,110,100. The results are updated accordingly.

danpovey commented:
@pswietojanski, don't forget about this!  I think the changes I asked for were quite small.

pswietojanski commented:
I will try to push another set of fixes this or next week (at latest). Thanks for the feedback Dan.

danpovey commented:
@pswietojanski are you still intending to work on this?

pswietojanski commented:
> @pswietojanski are you still intending to work on this?

Hi Dan. Apologies, I was changing jobs at that time, and somewhat overlooked and forgot about this. I guess this recipe is quite old already. I have some time next week (and spare resources), could you recommend which chain model is up to date so I can swap and rerun stuff so it is up to date? Is updating to mini_librispeech OK?

jtrmal commented:
@aarora8 have a look

pswietojanski commented:
> @aarora8 have a look

Hi @jtrmal and @aarora8 I commited fixes Dan asked for (removed dependency on irlstm, now all is done with srilm), and some other minor things. I rerun first stages of the recipe (data prep + gmm) to make sure these are comparable, and work properly.

1) There are still issues reported by CodeFactor in bash scripts (this hasn't been checked when original PR was made). Are these mandatory to resolve?
2) W.r.t chain, I am more than happy to merge the current models But if it's necessary, I may try to align these with what is currently SOTA model for AMI/s5b recipe.

Thanks both.

aarora8 commented:
Thanks, I will try to run it.

danpovey commented:
Thanks,  guys!  Ashish, make a reminder to get back to us even if you don't test it, I can consider merging it anyway.

aarora8 commented:
ok.

aarora8 commented:
For the GMM-HMM setup, I am getting a similar dev/eval WER (35.9/26.1) as in the RESULTS file.

pswietojanski commented:
I am also training ihm/mdm/sdm models to get more recent chain-based numbers comparable with AMI. I should commit these later today (so please do not merge at least until then).

pswietojanski commented:
Hi, so I pushed the recipes / results for more up-to-date FTDNN models. I am not planning to work on this PR anymore, unless you spot something that needs fixes.

danpovey commented:
Thanks!  Merged.
